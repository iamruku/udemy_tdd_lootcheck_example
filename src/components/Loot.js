import React, { Component } from 'react';
import { connect } from 'react-redux';
import { fetchBitcoin } from '../actions/actions';

export class Loot extends Component {
	componentDidMount() {
		this.props.fetchBitcoin();
	}

	computeBitcoin = () => {
		const { bitcoin } = this.props;

		if (Object.keys(bitcoin).length === 0) return '';

		return (
			this.props.balance / parseInt(bitcoin.bpi.USD.rate.replace(',', ''), 10)
		);
	};

	render() {
		return (
			<div>
				<h3>Bitcoin Balance: {this.computeBitcoin()}</h3>
			</div>
		);
	}
}

const mapStateToProps = ({ balance, bitcoin }) => ({ balance, bitcoin });

export default connect(
	mapStateToProps,
	{ fetchBitcoin }
)(Loot);
