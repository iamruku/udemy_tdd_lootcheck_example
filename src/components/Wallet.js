import React, { Component } from 'react';
import { connect } from 'react-redux';
import { depositBalance, withdrawBalance } from '../actions/actions';

export class Wallet extends Component {
	state = { inputBalance: 0 };

	deposit = () => {
		this.props.depositBalance(this.state.inputBalance);
	};

	withdraw = () => {
		this.props.withdrawBalance(this.state.inputBalance);
	};

	render() {
		return (
			<div>
				<h2 className="balance">Wallet Balance: {this.props.balance}</h2>
				<br />
				<label>Balance&nbsp;</label>
				<input
					type="text"
					className="inp-balance"
					onChange={e =>
						this.setState({ inputBalance: parseFloat(e.target.value) })
					}></input>
				<button className="btn-deposit" onClick={this.deposit}>
					Deposit
				</button>
				<button className="btn-withdraw" onClick={this.withdraw}>
					Withdraw
				</button>
			</div>
		);
	}
}

const mapStateToProps = ({ balance }) => ({ balance });

export default connect(
	mapStateToProps,
	{ depositBalance, withdrawBalance }
)(Wallet);
