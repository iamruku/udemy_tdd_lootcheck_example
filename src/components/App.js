import React from 'react';
import Wallet from './Wallet';
import Loot from './Loot';

export default () => (
	<div>
		<h2>Loot Check</h2>
		<hr></hr>
		<Wallet></Wallet>
		<Loot></Loot>
	</div>
);
