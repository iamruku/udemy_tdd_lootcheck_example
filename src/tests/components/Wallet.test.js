import React from 'react';
import { shallow } from 'enzyme';
import '../setupTests';
import { Wallet } from '../../components/Wallet';

describe('Wallet', () => {
	const _depositBalance = jest.fn();
	const _withdrawBalance = jest.fn();
	const props = {
		balance: 20,
		depositBalance: _depositBalance,
		withdrawBalance: _withdrawBalance
	};

	const wrapper = shallow(<Wallet {...props}></Wallet>);

	describe('initial dom content', () => {
		it('renders wrapper', () => {
			expect(wrapper).toMatchSnapshot();
		});

		it('displays the balance from props', () => {
			expect(wrapper.find('.balance').text()).toEqual(
				`Wallet Balance: ${props.balance}`
			);
		});

		it('contains an input to enter a balance amount', () => {
			expect(wrapper.find('.inp-balance').exists()).toBe(true);
		});

		it('contains a button to deposit the balance', () => {
			expect(wrapper.find('.btn-deposit').exists()).toBe(true);
		});

		it('contains a button to withdraw the balance', () => {
			expect(wrapper.find('.btn-withdraw').exists()).toBe(true);
		});
	});

	describe('initial state content', () => {
		it('should initialize `inputBalance` to 0', () => {
			expect(wrapper.state().inputBalance).toEqual(0);
		});
	});

	describe('change in input balance', () => {
		const balance = '25';

		beforeEach(() => {
			wrapper
				.find('.inp-balance')
				.simulate('change', { target: { value: balance } });
		});

		it('should update inputBalance prop in `state`', () => {
			expect(wrapper.state().inputBalance).toEqual(parseFloat(balance));
		});

		describe('user wants to deposit balance', () => {
			beforeEach(() => {
				wrapper.find('.btn-deposit').simulate('click');
			});

			it('dispatches the `depositBalance()` via props', () => {
				expect(_depositBalance).toHaveBeenCalledWith(parseFloat(balance));
			});
		});

		describe('user wants to withdraw balance', () => {
			beforeEach(() => {
				wrapper.find('.btn-withdraw').simulate('click');
			});

			it('dispatches the `withdrawBalance()` via props', () => {
				expect(_withdrawBalance).toHaveBeenCalledWith(parseFloat(balance));
			});
		});
	});
});
