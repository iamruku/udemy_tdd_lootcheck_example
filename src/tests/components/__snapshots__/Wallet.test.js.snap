// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Wallet initial dom content renders wrapper 1`] = `ShallowWrapper {}`;
